# Numerical Analysis Assignment #5

2018112749 전현승


## 1. Gradient Descent

In [2]:
import numpy as np

In [3]:
x = np.array([[2, 4, 6, 8], [0, 4, 2, 3], [1, 1, 1, 1]]).T
y = np.array([[81, 93, 91, 97]]).T
p = np.array([[0, 0, 0]]).T
J = -x.T

print(x)
print(x.shape)
print(y)
print(y.shape)
print(p)
print(p.shape)
print(J)
print(J.shape)

[[2 0 1]
 [4 4 1]
 [6 2 1]
 [8 3 1]]
(4, 3)
[[81]
 [93]
 [91]
 [97]]
(4, 1)
[[0]
 [0]
 [0]]
(3, 1)
[[-2 -4 -6 -8]
 [ 0 -4 -2 -3]
 [-1 -1 -1 -1]]
(3, 4)


In [4]:
lr = 0.01
iter = 1000

for i in range(iter):
  pred = np.matmul(x, p)
  error = (y - pred)
  p = p - lr * np.matmul(J, error)

error = y - np.matmul(x, p)
mse = np.matmul(error.T, error) / 4

print(mse)

[[0.07418457]]


## 2. Newton-Raphson

In [5]:
x = np.array([[2, 4, 6, 8], [0, 4, 2, 3], [1, 1, 1, 1]]).T
y = np.array([[81, 93, 91, 97]]).T
p = np.array([[0, 0, 0]]).T
J = -x.T
invH = np.linalg.inv(np.matmul(J, J.T))

print(x)
print(x.shape)
print(y)
print(y.shape)
print(p)
print(p.shape)
print(J)
print(J.shape)
print(invH)
print(invH.shape)

[[2 0 1]
 [4 4 1]
 [6 2 1]
 [8 3 1]]
(4, 3)
[[81]
 [93]
 [91]
 [97]]
(4, 1)
[[0]
 [0]
 [0]]
(3, 1)
[[-2 -4 -6 -8]
 [ 0 -4 -2 -3]
 [-1 -1 -1 -1]]
(3, 4)
[[ 0.06944444 -0.05555556 -0.22222222]
 [-0.05555556  0.15873016 -0.07936508]
 [-0.22222222 -0.07936508  1.53968254]]
(3, 3)


In [6]:
lr = 0.01
iter = 1000

for i in range(iter):
  pred = np.matmul(x, p)
  error = (y - pred)
  p = p - lr * np.matmul(np.matmul(invH, J), error)

error = y - np.matmul(x, p)
mse = np.matmul(error.T, error) / 4

print(mse)

[[0.0714439]]


## 3. Compare estimated parameters of two different methods.

1번 Gradient Descent 방법으로는 MSE가 0.07418457, 2번 Newton-Raphson 방법으로는 MSE가 0.0714439로 측정되었다. 결과만 보면 Gradient Descent 방법이 더 정확하지만, 여러 번 측정을 통해 결과를 비교해본다면 더 정밀한 비교가 가능할 것이다.